Add specialisations of `make_device_view` with a Queue argument

fwyzard opened the pull request:
#### PR description:

Add specialisations of `make_device_view` with a `Queue` argument instead of a `Device`.

These functions are only for convenience: their behaviour is identical to those taking a `Device`.


#### PR validation:

None.

fwyzard committed:
Add specialisations of make_device_view with a Queue argument

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"b9a0ab95706662e967199117e5027fcac7802c70":{"files":["HeterogeneousCore/AlpakaInterface/interface/memory.h"],"squashed":false,"time":1722229350}},"emoji":{},"last_seen_sha":"b9a0ab95706662e967199117e5027fcac7802c70","signatures":{}} -->

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45572/41068

fwyzard committed:
Simplify code using make_device_view

Simplify code using

    make_device_view(alpaka::getDev(queue), ...)

to use the new specialisation

    make_device_view(queue, ...)

.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45572/41069

cmsbuild commented:
Pull request #45572 was updated. @cmsbuild, @fwyzard, @jfernan2, @makortel, @mandrenguyen, @srimanob, @subirsarkar can you please check and sign again.

fwyzard commented:
please test

fwyzard commented:
@makortel this seems to be useful to reduce the calls to `alpaka::getDev(queue)`.
However, it has to be clear to the users that
```c++
make_device_view(queue, ...);
```
is just a shorthand for
```c++
make_device_view(alpaka::getDev(queue), ...);
```    
and does _not_ introduce any asynchronous behaviour.

Do you think this is OK, or potentially too confusing ?

cmsbuild commented:
-1

**Failed Tests**:   Build  HeaderConsistency  ClangBuild 
**Size**: This PR adds an extra 52KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-75d89d/40655/summary.html
**COMMIT**: 75cd6acd77f237fb8ec74ab23e2697b491d5d634
**CMSSW**: CMSSW_14_1_X_2024-07-28-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/45572/40655/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>Copying tmp/el8_amd64_gcc12/src/HeterogeneousTest/AlpakaWrapper/src/alpaka/HeterogeneousTestAlpakaWrapperROCmAsync/libHeterogeneousTestAlpakaWrapperROCmAsync_rocm.a to productstore area:
Copying tmp/el8_amd64_gcc12/src/RecoParticleFlow/PFRecHitProducer/plugins/RecoParticleFlowPFRecHitProducersPluginsPortableROCmAsync/libRecoParticleFlowPFRecHitProducersPluginsPortableROCmAsync_rocm.a to productstore area:
Copying tmp/el8_amd64_gcc12/src/DataFormats/TrackingRecHitSoA/test/Hits_testROCmAsync/libHits_testROCmAsync_rocm.a to productstore area:
cp: cannot stat 'tmp/el8_amd64_gcc12/src/DataFormats/TrackingRecHitSoA/test/Hits_testROCmAsync/libHits_testROCmAsync_rocm.a': No such file or directory
>> Deleted: tmp/el8_amd64_gcc12/src/DataFormats/TrackingRecHitSoA/test/Hits_testROCmAsync/libHits_testROCmAsync_rocm.a
gmake: *** [config/SCRAM/GMake/Makefile.rules:1864: tmp/el8_amd64_gcc12/src/DataFormats/TrackingRecHitSoA/test/Hits_testROCmAsync/libHits_testROCmAsync_rocm.a] Error 1
Copying tmp/el8_amd64_gcc12/src/DataFormats/SiPixelClusterSoA/test/Clusters_testROCmAsync/libClusters_testROCmAsync_rocm.a to productstore area:
Copying tmp/el8_amd64_gcc12/src/DataFormats/SiPixelDigiSoA/test/DigiErrors_testROCmAsync/libDigiErrors_testROCmAsync_rocm.a to productstore area:
Copying tmp/el8_amd64_gcc12/src/DataFormats/TrackSoA/test/TrackSoAHeterogeneousAlpaka_testROCmAsync/libTrackSoAHeterogeneousAlpaka_testROCmAsync_rocm.a to productstore area:
Copying tmp/el8_amd64_gcc12/src/DataFormats/VertexSoA/test/ZVertexSoA_testROCmAsync/libZVertexSoA_testROCmAsync_rocm.a to productstore area:
Copying tmp/el8_amd64_gcc12/src/RecoLocalCalo/HcalRecProducers/plugins/RecoLocalCaloHcalRecProducersPluginsPortableROCmAsync/libRecoLocalCaloHcalRecProducersPluginsPortableROCmAsync_rocm.a to productstore area:
</pre>

## Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Entering Package RecoLocalCalo/EcalRecProducers
>> Entering Package RecoLocalCalo/HGCalRecProducers
>> Entering Package RecoLocalTracker/SiPixelClusterizer
>> Entering Package RecoTracker/PixelSeeding
>> Compile sequence completed for CMSSW CMSSW_14_1_X_2024-07-28-2300
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 1
+ eval scram build outputlog '&&' '(python3' /data/cmsbld/jenkins/workspace/ib-run-pr-tests/cms-bot/buildLogAnalyzer.py --logDir /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_1_X_2024-07-28-2300/tmp/el8_amd64_gcc12/cache/log/src '||' 'true)'
++ scram build outputlog
>> Entering Package DataFormats/TrackingRecHitSoA
Entering library rule at DataFormats/TrackingRecHitSoA
>> Compiling  src/DataFormats/TrackingRecHitSoA/src/classes.cc
</pre>

makortel commented:
> Do you think this is OK, or potentially too confusing ?

Hmm. On a quick thought I think the _potential_ for misunderstanding is certainly there. I need to digest a bit more to reach a conclusion whether I'd think it would be too confusing or not. In any case documenting the behavior could help.

One possible direction to think about would be to add e.g. `async` postfix to all functions that may incur asynchronous behavior on some backend(s).

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_14_2_X. Please open a backport if it should also go in to CMSSW_14_1_X.

fwyzard commented:
> One possible direction to think about would be to add e.g. `async` postfix to all functions that may incur asynchronous behavior on some backend(s).

I think I would prefer to go the other way around, and add a `sync` postfix to the functions that block and synchronise - the rationale being that most functions taking a `queue` parameter _should_ be asynchronous.

However, how would you describe `make_device_view(queue, ...)` ?
It doesn't synchronise and block.
It's asynchronous in itself, but it creates a `View` that does not have asynchronous behaviour.
🤷🏻

fwyzard commented:
enable gpu

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45572/41600

cmsbuild commented:
Pull request #45572 was updated. @fwyzard, @jfernan2, @makortel, @mandrenguyen, @srimanob, @subirsarkar can you please check and sign again.

cmsbuild commented:
+1

**Size**: This PR adds an extra 28KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-75d89d/41228/summary.html
**COMMIT**: 569f2a90643c0c5667a34797d0d96d328bb6046f
**CMSSW**: CMSSW_14_2_X_2024-09-02-1100/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/45572/41228/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-02-1100+75d89d/64298/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 79 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-02-1100+75d89d/64298/validateJR.html)
 - DQMHistoTests: Total files compared: 45
 - DQMHistoTests: Total histograms compared: 3434837
 - DQMHistoTests: [Total failures: 4595](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-02-1100+75d89d/64298/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3430222
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 44 files compared)
 - Checked 197 log files, 168 edm output root files, 45 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-02-1100+75d89d/64298/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_2_X_2024-09-02-1100+75d89d/64300/validateJR.html)
 - DQMHistoTests: Total files compared: 6
 - DQMHistoTests: Total histograms compared: 37044
 - DQMHistoTests: [Total failures: 17](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_2_X_2024-09-02-1100+75d89d/64300/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 37027
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 5 files compared)
 - Checked 20 log files, 25 edm output root files, 6 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_2_X_2024-09-02-1100+75d89d/64300/triggerResults): no differences found

jfernan2 commented:
+1

makortel commented:
> However, how would you describe `make_device_view(queue, ...)` ?
> It doesn't synchronise and block.
> It's asynchronous in itself, but it creates a `View` that does not have asynchronous behaviour.

In a way both "synchronous" and "asynchronous" functions alter the state of the `queue`, from the point of view of the caller:
* a "sync" function (that e.g. calls `alpaka::wait(queue)`) guarantees the `queue` is in "synchronized state" (i.e. all memory accessed by the earlier operations can be used without further synchronization), regardless of the state of the `queue` before the function call
* a "async" function guarantees the `queue` is in "non-synchronized state" (i.e. any further access to the memory accessed also by the earlier operations must be synchronized either via `queue` or `alpaka::wait(queue)`), regardless of the state of the `queue` before the function call

From this point of view, the `make_device_view(queue, ...)` would be a "queue state preserving" function, i.e. the state of the `queue` after the `make_device_view()` function is whatever it was before the `make_device_view()` call.

But written all that, I fear making this point of view explicit in the function name would be more confusing, and therefore I think the present interface is probably good enough, as long as the behavior wrt. the `queue` is documented.

srimanob commented:
+Upgrade

fwyzard committed:
Document the memory-related helper types and functions

fwyzard commented:
+heterogeneous

fwyzard commented:
please test

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45572/41729

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45572/41729/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45572/41729/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45572/41731

cmsbuild commented:
Pull request #45572 was updated. @cmsbuild, @fwyzard, @jfernan2, @makortel, @mandrenguyen, @srimanob, @subirsarkar can you please check and sign again.

fwyzard commented:
please test

fwyzard commented:
+heterogeneous

jfernan2 commented:
+1

cmsbuild commented:
+1

**Size**: This PR adds an extra 36KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-75d89d/41446/summary.html
**COMMIT**: 85c2f00a51de5547372cbbdc8b0b1c05fcda9ee1
**CMSSW**: CMSSW_14_2_X_2024-09-10-1100/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/45572/41446/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 3 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-10-1100+75d89d/64489/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 14 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-10-1100+75d89d/64489/validateJR.html)
 - DQMHistoTests: Total files compared: 45
 - DQMHistoTests: Total histograms compared: 3437830
 - DQMHistoTests: [Total failures: 8](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-10-1100+75d89d/64489/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3437802
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 44 files compared)
 - Checked 197 log files, 168 edm output root files, 45 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-10-1100+75d89d/64489/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - You potentially [removed 36 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_2_X_2024-09-10-1100+75d89d/64490/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_2_X_2024-09-10-1100+75d89d/64490/validateJR.html)
 - DQMHistoTests: Total files compared: 7
 - DQMHistoTests: Total histograms compared: 53003
 - DQMHistoTests: [Total failures: 167](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_2_X_2024-09-10-1100+75d89d/64490/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 52836
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 6 files compared)
 - Checked 24 log files, 30 edm output root files, 7 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_2_X_2024-09-10-1100+75d89d/64490/triggerResults): no differences found

makortel commented on HeterogeneousCore/AlpakaInterface/README.md:
```diff
@@ -4,3 +4,310 @@ This package only depends on the `alpaka` header-only external library, and
+On most systems accessing this memory from the device is significantly slower
+than using device global memory, so it is not adviced to use these buffers
+directly in device kernels. A common use case for device-mapped memory is as a
+staging area for copies to and from the devices.
```
Would it be worth of mentioning somewhere here that the following cases require device-host synchronization (since these pitfalls do not apply to the device buffers)?

* after `alpaka::memcpy(queue, ...)` from host to device (or device kernels that would access the host memory directly, even if that was discouraged above), the source host buffer can be written into again only after a synchronization
* after `alpaka::mempcy(queue, ...)` from device to host, the destination host buffer can be accessed (read or write) only after a synchronization

While these are sort of obvious, I feel mentioning them explicitly could be useful for newcomers

fwyzard replied:
Good points, thanks for mentioning them.

I am also thinking that a `host_buffer` with pinned memory _could_ be used for communication between the GPU and CPU (_e.g._ the GPU could write there a size, and the CPU could later use it to allocate a buffer sized accordingly); however this is quite advanced and would require some _ad-hoc_ host-device synchronisation, so maybe this would be outside the scope of this documentation ?

makortel replied:
> I am also thinking that a `host_buffer` with pinned memory _could_ be used for communication between the GPU and CPU (_e.g._ the GPU could write there a size, and the CPU could later use it to allocate a buffer sized accordingly); however this is quite advanced and would require some _ad-hoc_ host-device synchronisation, so maybe this would be outside the scope of this documentation ?

Yeah, I feel this pattern is something that should be tried out first before documenting it.

makortel commented on HeterogeneousCore/AlpakaInterface/README.md:
```diff
@@ -4,3 +4,310 @@ This package only depends on the `alpaka` header-only external library, and
+These type aliases simplify the use of zero-dimensional (scalars) and
+one-dimensional (array) alpaka device views:
+
+  - `device_view<T>` (where `T` is not an array)
```
Should these have the `Device` template argument? (like `device_buffer` has)

```suggestion
  - `device_view<Device, T>` (where `T` is not an array)
```

fwyzard replied:
I think so, yes.
Thanks, I will add them.

makortel reviewed (commented):
Thanks @fwyzard for the extensive documentation. I have a few suggestions (possibly for a subsequent PR if not here).

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45572/41756

cmsbuild commented:
Pull request #45572 was updated. @cmsbuild, @fwyzard, @jfernan2, @makortel, @mandrenguyen, @srimanob, @subirsarkar can you please check and sign again.

fwyzard commented:
Updated the documentation and rebased on top of a more recent release.

fwyzard commented:
please test

fwyzard commented:
+heterogeneous

jfernan2 commented:
+1

fwyzard commented:
@cms-sw/upgrade-l2 could you sign this, please ?

Alternatively, @cms-sw/orp-l2 could you bypass the upgrade signature and merge this ?
It's now been pending for two weeks.

srimanob commented:
+Upgrade

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @rappoccio, @sextonkennedy, @antoniovilela, @mandrenguyen (and backports should be raised in the release meeting by the corresponding L2)

mandrenguyen commented:
+1
